pkg/model/model.go

```diff
@@ -0,0 +1,295 @@
+	//	result = multierror.Append(result, errors.New("icon data must be set if icon is defined"))
+	//}
+	//if len(i.MediaType) == 0 {
+	//	result = multierror.Append(result, errors.New("icon mediatype must be set if icon is defined"))
+	//}
+	//if len(i.Data) > 0 {
+	//	if err := i.validateData(); err != nil {
+	//		result = multierror.Append(result, err)
+	//	}
+	//}
+	return result.ErrorOrNil()
+}
+
+func (i *Icon) validateData() error {
+	if !filetype.IsImage(i.Data) {
+		return errors.New("icon data is not an image")
+	}
+	t, err := filetype.Match(i.Data)
+	if err != nil {
+		return err
+	}
+	if t.MIME.Value != i.MediaType {
+		return fmt.Errorf("icon media type %q does not match detected media type %q", i.MediaType, t.MIME.Value)
+	}
+	return nil
+}
+
+type Channel struct {
+	Package *Package
+	Name    string
+	Bundles map[string]*Bundle
+}
+
+// TODO(joelanford): This function determines the channel head by finding the bundle that has 0
+//   incoming edges, based on replaces and skips. It also expects to find exactly one such bundle.
+//   Is this the correct algorithm?
+func (c Channel) Head() (*Bundle, error) {
+	incoming := map[string]int{}
+	for _, b := range c.Bundles {
+		if b.Replaces != "" {
+			incoming[b.Replaces]++
+		}
+		for _, skip := range b.Skips {
+			incoming[skip]++
+		}
+	}
+	var heads []*Bundle
+	for _, b := range c.Bundles {
+		if _, ok := incoming[b.Name]; !ok {
+			heads = append(heads, b)
+		}
+	}
+	if len(heads) == 0 {
+		return nil, fmt.Errorf("no channel head found in graph")
+	}
+	if len(heads) > 1 {
+		var headNames []string
+		for _, head := range heads {
+			headNames = append(headNames, head.Name)
+		}
+		return nil, fmt.Errorf("multiple channel heads found in graph: %s", strings.Join(headNames, ", "))
+	}
+	return heads[0], nil
+}
+
+func (c *Channel) Validate() error {
+	var result *multierror.Error
+	if c.Name == "" {
+		result = multierror.Append(result, errors.New("channel name must not be empty"))
+	}
+
+	if c.Package == nil {
+		result = multierror.Append(result, errors.New("package must be set"))
+	}
+
+	if len(c.Bundles) == 0 {
+		result = multierror.Append(result, fmt.Errorf("channel must contain at least one bundle"))
+	}
+
+	if len(c.Bundles) > 0 {
+		if _, err := c.Head(); err != nil {
+			result = multierror.Append(result, err)
+		}
+	}
+
+	for name, b := range c.Bundles {
+		if name != b.Name {
+			result = multierror.Append(result, fmt.Errorf("bundle key %q does not match bundle name %q", name, b.Name))
+		}
+		if err := b.Validate(); err != nil {
+			result = multierror.Append(result, fmt.Errorf("invalid bundle %q: %v", b.Name, err))
+		}
+		if b.Channel != c {
+			result = multierror.Append(result, fmt.Errorf("bundle %q not correctly linked to parent channel", b.Name))
+		}
+	}
+	return result.ErrorOrNil()
+}
+
+type Bundle struct {
+	Package       *Package
+	Channel       *Channel
+	Name          string
+	Image         string
+	Replaces      string
+	Skips         []string
```

Comment:
I seem to be missing something, did we decided on not storing the skipRange for the bundle?

Reply:
There's a property for skipRange.